static/game/levelpresets.js

```diff
@@ -0,0 +1,180 @@
+ * @property {number} powerScalar
+ * @property {number} randomPowerAddend
+ * @property {number} tilesPerChance
+ *
+ * @typedef DimensionSettings
+ * @property {number} roomWidth
+ * @property {number} roomHeight
+ * @property {number} blockWidth
+ * @property {number} blockHeight
+ *
+ * @typedef GameModeSettings
+ * @property {{minutes: number, seconds: number}} timeLimit
+ *
+ * @typedef SettingsGroup
+ * @property {TerrainSettings} terrain
+ * @property {SpawnSettings} spawn
+ * @property {DimensionSettings} dimensions
+ * @property {GameModeSettings} gameMode
+ */
+
+const DEFAULT_BLOCK_WIDTH = 60;
+const DEFAULT_BLOCK_HEIGHT = 60;
+
+/** @type {Object<string, TerrainSettings>} */
+export const terrainSettings = {
+  original: {
+    rules: caveRules,
+    density: 0.5,
+    generations: 20
+  }
+};
+
+/** @type {Object<string, SpawnSettings>} */
+export const spawnSettings = {
+  original: {
+    chance: 0.025,
+    densityDistanceLo: 1000,
+    densityDistanceHi: 5000,
+    powerDistanceLo: 1000,
+    powerDistanceHi: 5000,
+    powerScalar: 3,
+    randomPowerAddend: 3,
+    tilesPerChance: 280
+  }
+};
+
+/** @type {Object<string, DimensionSettings>} */
+export const dimensionsSettings = {
+  original: {
+    roomWidth: 256,
+    roomHeight: 144,
+    blockWidth: DEFAULT_BLOCK_WIDTH,
+    blockHeight: DEFAULT_BLOCK_HEIGHT
+  }
+};
+
+/** @type {Object<string, GameModeSettings>} */
+export const gameModeSettings = {
+  original: {
+    timeLimit: { minutes: 5, seconds: 0 }
+  }
+};
+
+/** @type {Object<string, SettingsGroup>} */
+export const settingsGroups = {
+  original: {
+    terrain: terrainSettings.original,
+    spawn: spawnSettings.original,
+    dimensions: dimensionsSettings.original,
+    gameMode: gameModeSettings.original
+  }
+};
+
+/**
+ * generate terrain, spawn enemies, spawn powerups and set game mode based
+ * on a group of settings
+ * @param {SettingsGroup} group
+ */
+export function startLevelFromSettings(group) {
+  const timedisplay = new TimeDisplay(
+    new Vector(0, 200 - 32),
+    100 *
+      (group.gameMode.timeLimit.minutes * 60 + group.gameMode.timeLimit.seconds)
+  );
+  addToGui("timedisplay", timedisplay);
+
+  const board = getGrid(
+    group.dimensions.roomWidth,
+    group.dimensions.roomHeight,
+    group.terrain.rules,
+    EdgesEnum.alive,
+    group.terrain.density,
+    group.terrain.generations
+  );
+
+  const color = hsl(randomInt(360));
+
+  setGameDrawFunc(() => {
+    drawBoard(
+      board,
+      group.dimensions.blockWidth,
+      group.dimensions.blockHeight,
+      color
+    );
+  });
+
+  let emptySpaces = shuffle(
+    getEmptySpaces(
+      board,
+      10,
+      group.dimensions.blockWidth,
+      group.dimensions.blockHeight
+    )
+  );
+
+  // TODO change the hero spawning
+  const hero = new Hero(
+    new Vector(0, 0).add(
+      new Vector(
+        group.dimensions.blockWidth / 2,
+        group.dimensions.blockHeight / 2
+      ).add(emptySpaces[0])
```

Comment:
this used to choose a random empty space, right?
wouldn't they all be in the top right corner now?

Reply:
> this used to choose a random empty space, right?
> wouldn't they all be in the top right corner now?

the list is shuffled